### DISCUSSION TITLE
getting error: Variable 0p_FoldList_ is never assigned

### DISCUSSION THREAD

**Discussion by reitcircles** [OpShin/opshin#326] [Q&A]
Following is the datum structure in opshin:
```@dataclass()
  class Policydatum(PlutusData):
    oref: TxOutRef
    name_usr_nft: bytes
    name_ref_nft: bytes`

```

```
And trying to just contruct dynamically using the following code:

        oref = TxOutRef(
            id=TxId(bytes.fromhex(uref["txId"])),
            idx=uref['txIndex']
        )
        tn_usr_bytes = bytes(params['tokenName_usr'], encoding="utf-8")  #convert to bytes                                                                                                                  
        tn_ref_bytes = bytes(params['tokenName_ref'], encoding="utf-8")  #convert to bytes                                                                                                                  

	policy_datum = Policydatum(oref, tn_usr_bytes, tn_ref_bytes)
```

It was working earlier but now this seems to be a problem

opshin: Version: 0.18.1
pycardano: Version: 0.9.0

**Comment by nielstron** [comment]
Can you please report the full error message and ideally the surrounding code of the code invocation in a minimal example?

**Reply by reitcircles** [reply]
look above ...

```
def build_mint_policy_script(self, params={}):
        """                                                                                                                                                                                                 
        params should be the same set  of inputs as validator requires.                                                                                                                                     
        This is to build the opshin code to a cbor (plutus) format, using a set of predefined inputs for validator                                                                                          
                                                                                                                                                                                                            
        {                                                                                                                                                                                                   
          output_ref:{                                                                                                                                                                                      
                txId: "",                                                                                                                                                                                   
                txIndex: ""                                                                                                                                                                                 
          },                                                                                                                                                                                                
          tokenName_usr: "(100)USER",                                                                                                                                                                       
          tokenName_ref: "(222)REFERENCE"                                                                                                                                                                   
        }                                                                                                                                                                                                   
        """

        uref = params['output_ref']

        print(uref)

        oref = TxOutRef(
            id=TxId(bytes.fromhex(uref["txId"])),
            idx=uref['txIndex']
        )
        tn_usr_bytes = bytes(params['tokenName_usr'], encoding="utf-8")  #convert to bytes                                                                                                                  
        tn_ref_bytes = bytes(params['tokenName_ref'], encoding="utf-8")  #convert to bytes                                                                                                                  


        policy_datum = Policydatum(oref, tn_usr_bytes, tn_ref_bytes)
        policy_datum_cbor_hex = policy_datum.to_cbor_hex()

        #Now can we inject the above datum to the policy mint validator ? Let us try.It returns of type:PlutusV2Script                                                                                      
        script_path = os.path.join(os.path.dirname(__file__), "../sc/opshin/nft_policy.py")
        mint_script = build(script_path, policy_datum)
        policy_id = plutus_script_hash(mint_script)

        print(f"{policy_id}")
        return (mint_script, policy_id)
```

**Reply by nielstron** [reply]
Thanks for providing these details :) which contract are you executing? Is this a standard example?

**Reply by nielstron** [reply]
Also, does this error occur during building (i.e. running the `build` command) or after submitting the transaction?

**Reply by nielstron** [reply]
And finally, please check if opshin version 0.19.0 fixes this bug and/or the `dev` branch of OpShin fixes this bug.

**Reply by reitcircles** [reply]
yes ...basically we are creating cip68 NFT where we are basically consuming a unique utxo and creating a policyid out of it. Hence this exercice

**Reply by reitcircles** [reply]
I tried to hardcode the versions ...but gave up because of interdependecy of version between opshin, pycardano and many other packages. And this led to this set of versions, where at least i am now getting debug messages, while earlier it was 500 without any message ...(and at the last moment when i tried to sign the tx)

**Reply by reitcircles** [reply]
just check the code above ....

**Comment by reitcircles** [comment]
Do you have any clue ...that i can use to debug this topic ?

**Reply by nielstron** [reply]
This looks to me like an issue on the level of plyto. You might want to try and upgrade the `plyto` version used. However pycardano 0.10.0 and opshin 0.19.0 should also work.

**Reply by nielstron** [reply]
I never encountered this bug myself and I am regularly compiling and executing a number of contracts. If you share the smart contract code I can try to reproduce the error.

**Comment by reitcircles** [comment]
```

from opshin.prelude import *

@dataclass()
class Policydatum(PlutusData):
    CONSTR_ID = 1
    oref: TxOutRef
    name_usr_nft: bytes
    name_ref_nft: bytes

def has_utxo( oref: TxOutRef, context: ScriptContext) -> bool:
    return any([oref == i.out_ref for i in context.tx_info.inputs])


def validator( datum: Policydatum, redeemer: None, context: ScriptContext) -> None:
    print("ENTERED VALIDATOR")
    assert has_utxo(datum.oref, context), "UTxO not consumed"

```

**Reply by reitcircles** [reply]
ok. Opshin 0.19.0 has solved that problem, just precipating in another kind of error:
`{'error': 'Internal Server Error', 'message': 'An unexpected response was received from the backend.', 'status_code': 500}`

**Reply by reitcircles** [reply]
Now how to debug this without any explanation of error.

**Reply by reitcircles** [reply]
Is there any extra level of verbosity so we can check the actual error in the backend ?

**Reply by reitcircles** [reply]
Here is next problem of silent failing of opshin:
'''
TransactionBuilder(context=<pycardano.backend.blockfrost.BlockFrostChainContext object at 0x7f22f9bad130>, utxo_selectors=[<pycardano.coinselection.RandomImproveMultiAsset object at 0x7f22f3dc4940>, <pycardano.coinselection.LargestFirstSelector object at 0x7f22f3dc1b20>], execution_memory_buffer=0.2, execution_step_buffer=0.2, ttl=None, validity_start=None, auxiliary_data=AuxiliaryData(data=AlonzoMetadata(metadata={721: {'4ba551fde963349c45fd4f90fab01361853c7db9f2486383ada82969': {AssetName(b'(222)rcNFT'): {'tokenType': 'role', 'description': 'Role NFT from reitcircles.com', 'image': 'QmNmrtWycJARzQu7qWwVwB4azurBmCR2oSp3d9dtLdusCQ'}}}}, native_scripts=None, plutus_scripts=None)), native_scripts=None, mint={ScriptHash(hex='4ba551fde963349c45fd4f90fab01361853c7db9f2486383ada82969'): {AssetName(b'(100)rcNFT'): 1, AssetName(b'(222)rcNFT'): 1}}, required_signers=None, collaterals=[{'input': {
  'index': 1,
  'transaction_id': TransactionId(hex='c44e1113201927c3d079a98add29bb5a2f222e0889691291ed1433bc584d6d89'),
},
 'output': {
  'address': addr_test1qzh9j2qa7dny2rpjvgedvjmufaqndhgjejkmsmaatsdeufgd74hhzeetur4qr9v9e2mce0uxlzqelhuy9dq2027dh78ssquj4k,
  'amount': {'coin': 5000000, 'multi_asset': {}},
  'datum': None,
  'datum_hash': None,
  'post_alonzo': False,
  'script': None,
}}], certificates=None, withdrawals=None, reference_inputs=set(), _inputs=[{'input': {
  'index': 2,
  'transaction_id': TransactionId(hex='6fc171064584f0f1d95416f2031614d7de5c6da746e030cc19aad017b86a2bd7'),
},
 'output': {
  'address': addr_test1qzylgj0pzzm722x3vkp6nde7h5cuzt2m4g9c69n5dj0w5rqd74hhzeetur4qr9v9e2mce0uxlzqelhuy9dq2027dh78sgh34ld,
  'amount': {
    'coin': 874077053,
    'multi_asset': {ScriptHash(hex='fb23f71797eb9ed59172d89bba6a69ae97f9085d213033d9b9bb83c5'): {AssetName(b'sampleDao.001'): 8}},
  },
  'datum': None,
  'datum_hash': None,
  'post_alonzo': False,
  'script': None,
}}], _potential_inputs=[], _excluded_inputs=[], _input_addresses=[], _outputs=[{
  'address': addr_test1wrwnpedpcagj6nvp30r5s35awjat0ehrrzf9c6q5xhsgrjqx0khru,
  'amount': {
    'coin': 4000000,
    'multi_asset': {ScriptHash(hex='4ba551fde963349c45fd4f90fab01361853c7db9f2486383ada82969'): {AssetName(b'(100)rcNFT'): 1}},
  },
  'datum': RoleDatum(roleName=b'general', subscriptionLevel=b'Tier1', expiryDate=1688064662, pubkeyhash=b'\x89\xf4I\xe1\x10\xb7\xe5(\xd1e\x83\xa9\xb7>\xbd1\xc1-[\xaa\x0b\x8d\x16tl\x9e\xea\x0c'),
  'datum_hash': None,
  'post_alonzo': False,
  'script': None,
}, {
  'address': addr_test1qrwcxmlx56njze57fjypt2f6q2ctpv2tvx2gf2mf866ajsasqj43gyduezd0crpju8l9ps4fuu94z0rp86t00c4m3nzs8yf0nd,
  'amount': {
    'coin': 4000000,
    'multi_asset': {ScriptHash(hex='4ba551fde963349c45fd4f90fab01361853c7db9f2486383ada82969'): {AssetName(b'(222)rcNFT'): 1}},
  },
  'datum': None,
  'datum_hash': None,
  'post_alonzo': False,
  'script': None,
}], _fee=0, _datums={}, _collateral_return=None, _total_collateral=None, _inputs_to_redeemers={}, _minting_script_to_redeemers=[(b'Y\x02\xa0\x01\x00\x003############""2222222222222223t\xa9\x00\x01\xbb\x14\x98\xc8\xc8\xc8\xcc\xcc\xd4\x01\x00\x14\x02\x04\x00\xc4\x00\x84\x00@D\x92`\x11\x12""S3W4d`< \x02ddfj\x00\xe0\x12 \x04 \x02\xa0\x04`.\xa0\x06)0\x991\x9a\xb9\xc4\x91\x11UTxO not consumed\x00I\x84\x88\x8c\x8d@\x10@\x04\xcc\xc0`\xc0L\xc0Q@\x04\x8c\x8c\xc0e@\x10\xc0Y@\x04@\x04R\xf5\xa0 \x02FFf\x02\xc0\x02Df\x03\xa0\x04 \x02\x94\x14\x00D\xc9\x8c\xd5\xce$\x81\x10NameError: ~bool\x00I\x84\xc9\x8c\xd5\xce$\x81\x14NameError: validator\x00I\x84\xc9\x8c\xd5\xce$\x81\x13NameError: redeemer\x00I\x84\xc9\x8c\xd5\xce$\x81\x0fNameError: oref\x00I\x84\xc9\x8c\xd5\xce$\x81\x0cNameError: i\x00I\x84\xc9\x8c\xd5\xce$\x81\x13NameError: has_utxo\x00I\x84\xc9\x8c\xd5\xce$\x81\x10NameError: datum\x00I\x84\xc9\x8c\xd5\xce$\x81\x12NameError: context\x00I\x84\xc9\x8c\xd5\xce$\x81\x12NameError: context\x00I\x84\xc9\x8c\xd5\xce$\x81\x0eNameError: any\x00I\x80\x08\x00\x88\xcc\xc0\x10\xdda\x80A\x808\x00\x90\x00\x89\xbb\x14\x98\x8c\x01\xcc\x01\x80\x04\x01\x88\x88\xcc\x8c\x02\xc8\x94\xcc\xd5\\\xf8\x00\x8a\x80 \x99\xab\xa00\x035t \x02f\x00@\x04j\xe8\x80\x04\x00\x80\x0c\x88\x8c\xcc\x8c\x02\x88\x89L\xcdU\xcf\x80\x10\x80\x08\x99\x98\x01\x80\x19\xab\xa2\x00#0\x04\x00\x13WB\x00@\x04\x00`\x02F@\x02\xa0\x02F\xaa\xe7\x8d\xd5\x00\t\x19)\x99\xaa\xb9\xf0\x01\x13&3W8\x92\x01\nIndexError\x00I\x84\xd5\xd0\x80\x08\x00\xba\xf2%35s@\x04)DT\x00H\xc8\xc0\x04\x00@\x050\x01E\xd8z\x9f\xd8y\x9f\xd8y\x9fX o\xc1q\x06E\x84\xf0\xf1\xd9T\x16\xf2\x03\x16\x14\xd7\xde\\m\xa7F\xe00\xcc\x19\xaa\xd0\x17\xb8j+\xd7\xff\x02\xffJ(222)rcNFTJ(100)rcNFT\xff\x00\x01', {
  'data': {},
  'ex_units': {'mem': 0, 'steps': 0},
  'tag': {'__objclass__': <enum 'RedeemerTag'>, '_name_': 'MINT', '_value_': 1},
})], _inputs_to_scripts={}, _reference_scripts=[], _should_estimate_execution_units=True)
84ac00818258206fc171064584f0f1d95416f2031614d7de5c6da746e030cc19aad017b86a2bd7020183a300581d70dd30e5a1c7512d4d818bc748469d74bab7e6e318925c681435e081c801821a003d0900a1581c4ba551fde963349c45fd4f90fab01361853c7db9f2486383ada82969a14a283130302972634e465401028201d8185835d87a9f4767656e6572616c4554696572311a649dd296581c89f449e110b7e528d16583a9b73ebd31c12d5baa0b8d16746c9eea0cff82583900dd836fe6a6a721669e4c8815a93a02b0b0b14b619484ab693eb5d943b004ab1411bcc89afc0c32e1fe50c2a9e70b513c613e96f7e2bb8cc5821a003d0900a1581c4ba551fde963349c45fd4f90fab01361853c7db9f2486383ada82969a14a283232322972634e4654018258390089f449e110b7e528d16583a9b73ebd31c12d5baa0b8d16746c9eea0c0df56f71672be0ea019585cab78cbf86f8819fdf842b40a7abcdbf8f821a339bb268a1581cfb23f71797eb9ed59172d89bba6a69ae97f9085d213033d9b9bb83c5a14d73616d706c6544616f2e30303108021a00039715031a0263dab407582001c29487611a0364ff49594379380a0dabea257c717a397a39267c419d418efc081a0263afbc09a1581c4ba551fde963349c45fd4f90fab01361853c7db9f2486383ada82969a24a283130302972634e4654014a283232322972634e4654010b5820a503d63deba6b3c1b98af96b0b1ad11a89837493cc7e8eff5dd444fb8ddb231b0d81825820c44e1113201927c3d079a98add29bb5a2f222e0889691291ed1433bc584d6d89010e82581cae59281df366450c326232d64b7c4f4136dd12ccadb86fbd5c1b9e25581c89f449e110b7e528d16583a9b73ebd31c12d5baa0b8d16746c9eea0c108258390089f449e110b7e528d16583a9b73ebd31c12d5baa0b8d16746c9eea0c0df56f71672be0ea019585cab78cbf86f8819fdf842b40a7abcdbf8f1a00153f01111a00370c3fa300828258205797dc2cc919dfec0bb849551ebdf30d96e5cbe0f33f734a87fe826db30f7ef95840577ccb5b487b64e396b0976c6f71558e52e44ad254db7d06dfb79843e5441a5d763dd42adcf5e8805d70373722ebbce62a58e3f30dd4560b9a898b8ceeab6a038258205797dc2cc919dfec0bb849551ebdf30d96e5cbe0f33f734a87fe826db30f7ef95840577ccb5b487b64e396b0976c6f71558e52e44ad254db7d06dfb79843e5441a5d763dd42adcf5e8805d70373722ebbce62a58e3f30dd4560b9a898b8ceeab6a0306815902a35902a00100003323232323232323232323232322223232323232323232323232323232323374a90001bb1498c8c8c8ccccd4010014020400c400840040449260111222225333573464603c20026464666a00e01220042002a004602ea006293099319ab9c491115554784f206e6f7420636f6e73756d6564004984888c8d40104004ccc060c04cc05140048c8cc0654010c0594004400452f5a0200246466602c002446603a004200294140044c98cd5ce2481104e616d654572726f723a207e626f6f6c004984c98cd5ce2481144e616d654572726f723a2076616c696461746f72004984c98cd5ce2481134e616d654572726f723a2072656465656d6572004984c98cd5ce24810f4e616d654572726f723a206f726566004984c98cd5ce24810c4e616d654572726f723a2069004984c98cd5ce2481134e616d654572726f723a206861735f7574786f004984c98cd5ce2481104e616d654572726f723a20646174756d004984c98cd5ce2481124e616d654572726f723a20636f6e74657874004984c98cd5ce2481124e616d654572726f723a20636f6e74657874004984c98cd5ce24810e4e616d654572726f723a20616e79004980080088ccc010dd618041803800900089bb14988c01cc018004018888cc8c02c894ccd55cf8008a802099aba0300335742002660040046ae8800400800c888ccc8c0288894ccd55cf801080089998018019aba20023300400135742004004006002464002a00246aae78dd50009192999aab9f001132633573892010a496e6465784572726f72004984d5d0800800baf22533357340042944540048c8c004004005300145d87a9fd8799fd8799f58206fc171064584f0f1d95416f2031614d7de5c6da746e030cc19aad017b86a2bd7ff02ff4a283232322972634e46544a283130302972634e4654ff00010581840100d87980820000f5d90103a100a11902d1a178383462613535316664653936333334396334356664346639306661623031333631383533633764623966323438363338336164613832393639a14a283232322972634e4654a369746f6b656e5479706564726f6c656b6465736372697074696f6e781d526f6c65204e46542066726f6d2072656974636972636c65732e636f6d65696d616765782e516d4e6d72745779634a41527a51753771577756774234617a7572426d4352326f537033643964744c6475734351
--------------------------/tmp/tmpswlbuy3p---------------------------------------
{'error': 'Internal Server Error', 'message': 'An unexpected response was received from the backend.', 'status_code': 500}

'''

**Reply by nielstron** [reply]
Sounds like you are using the Blockfrost chain backend which does not give very detailed error messages. You can try using Ogmios instead which has very good error messages (you can try it for free using I.e. Demeter.run)

**Reply by reitcircles** [reply]
Do you know which version of ogmios works...Since i am SPO local nodes are running. Ogmios latest version 6.0.2 gives the following error when querying on preview network. Also earlier version of ogmios was disconnecting and hence was forced to install the latest one.
`
pycardano.exception.TransactionFailedException: Ogmios ran into an error. Reponse: {"jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid request: field 'method' must be present and be a string."},"id":null}`

And then i go into one version lower of ogmios v6.0.1 and then comes error: ` Error pydantic.error_wrappers.ValidationError: 1 validation error for BlockPraos
issuer -> operationalCertificate -> sigma
  extra fields not permitted (type=value_error.extra)`

**Reply by nielstron** [reply]
The latest version of Ogmios that pycardano support as of now is 5.x

**Reply by nilaysaha** [reply]
would you have a code piece to help me connect to the demeter.run instance. ? Some example. thanks. Local instances require more work as i need to downgrade the cardano-node before using the 5.x version of ogmios !

**Reply by nilaysaha** [reply]
I would strongly suggest, that a stable version of different packages is published when you push new releases of pycardano (so dependencies that are not conflicting with one another). This currently is a big problem as i install one version of pycardano, that conflicts with another version of opshin and that in turn has another conflict etc..very annoying frankly.

**Reply by nielstron** [reply]
> would you have a code piece to help me connect to the demeter.run instance. ?

Basically you will want to set up Ogmios in a Demeter.run project and expose the websockrt interface publicly, then connect to the exposed URL. An example is in the [OpShin-pioneer-program](https://github.com/OpShin/opshin-pioneer-program?tab=readme-ov-file#quick-setup).

**Reply by nielstron** [reply]
> I would strongly suggest, that a stable version of different packages is published when you push new releases of pycardano

Thank you for the suggestion. Please note

1) I do not push releases of PyCardano, it is a seperate project
2) The cardano-node 8.x release has broken a lot of tooling and we are doing our best to keep up given limited resources (this project and pycardano is entirely community-built)
3) I can highly recommend using `poetry` for python projects to handle dependencies, the projects clearly define which packages they are compatible with and are tested contiuously that the required configuration is feasible

**Reply by nilaysaha** [reply]
the connection gets reset when i try to sign using the context of demeter.run (`Connection to remote host was lost.`) ....Now kind of giving up .... just because we have no debug information ...

**Reply by nilaysaha** [reply]
Hi Niels. Could you pls check your email. I have sent some request. thanks.

**Reply by reitcircles** [reply]
it works again for some reason ...do not ask me why ...
